src/d3a/models/market/two_sided.py

```diff
@@ -257,35 +257,57 @@ def accept_bid_offer_pair(self, bid, offer, clearing_rate, trade_bid_info, selec
     def match_recommendations(
             self, recommendations: List[BidOfferMatch.serializable_dict]) -> None:
         """Match a list of bid/offer pairs, create trades and residual offers/bids."""
-        if recommendations is None:
-            return
-        for index, recommended_pair in enumerate(recommendations):
+
+        while recommendations:
+            recommended_pair = recommendations.pop()
             selected_energy = recommended_pair["selected_energy"]
-            market_offer = self.offers.get(recommended_pair["offer"]["id"])
-            market_bid = self.bids.get(recommended_pair["bid"]["id"])
-
-            if not market_offer:
-                raise OfferNotFoundException
-            if not market_bid:
-                raise BidNotFoundException
-
-            original_bid_rate = market_bid.original_bid_price / market_bid.energy
-            trade_bid_info = TradeBidOfferInfo(
-                original_bid_rate=original_bid_rate,
-                propagated_bid_rate=market_bid.price/market_bid.energy,
-                original_offer_rate=market_offer.original_offer_price/market_offer.energy,
-                propagated_offer_rate=market_offer.price/market_offer.energy,
-                trade_rate=original_bid_rate)
-
-            bid_trade, trade = self.accept_bid_offer_pair(
-                market_bid, market_offer, recommended_pair["trade_rate"],
-                trade_bid_info, selected_energy
-            )
-
-            if trade.residual is not None or bid_trade.residual is not None:
-                recommendations = self._replace_offers_bids_with_residual_in_matching_list(
-                    recommendations, index+1, trade, bid_trade
+            clearing_rate = recommended_pair["trade_rate"]
+            market_offers = [
+                self.offers.get(offer["id"]) for offer in recommended_pair["offers"]]
+            market_bids = [self.bids.get(bid["id"]) for bid in recommended_pair["bids"]]
+
+            if not all(list(market_offers)):
```

Comment:
Why is this being forced to `list`?

Reply:
I have originally made them as iterable and changed this later on, I'll update this line